### PR TITLE
Toast on CustomPanels

### PR DESCRIPTION
**Pull request by pjfitzgibbons**
### Description
Apply OSD GlobalToast to CustomPanels

### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/custom_panels/custom_panel_table.tsx [hunk 1]
```diff
@@ -222,17 +230,16 @@ export const CustomPanelTable = ({
         'Duplicate Dashboard',
         'Cancel',
         'Duplicate',
-        selectedCustomPanels[0].title + ' (copy)',
+        selectedCustomPanels[0].title + ' (copy)x',
```
Comment (ps48) [inline public/components/custom_panels/custom_panel_table.tsx:233]: We can remove the `x` typo here.

File: public/components/common/toast/index.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { ObservabilityAppServices } from '../../../../common/types/shared';
+
+type Color = 'success' | 'primary' | 'warning' | 'danger' | undefined;
+
+export const useToast = () => {
+  const {
+    services: { toasts },
+  } = useOpenSearchDashboards<ObservabilityAppServices>();
+
+  const setToast = (title: string, color: Color = 'success', text?, side?: string) => {
```
Comment (ps48) [inline public/components/common/toast/index.tsx:13]: We can remove the side parameter, if not being used.

File: public/components/custom_panels/custom_panel_table.tsx [hunk 3]
```diff
@@ -77,39 +78,32 @@ interface Props {
...
-  cloneCustomPanel,
-  deleteCustomPanelList,
   addSamplePanels,
 }: Props) => {
   const customPanels = useSelector<CustomPanelType[]>(selectPanelList);
   const [isModalVisible, setIsModalVisible] = useState(false); // Modal Toggle
   const [modalLayout, setModalLayout] = useState(<EuiOverlayMask />); // Modal Layout
   const [isActionsPopoverOpen, setIsActionsPopoverOpen] = useState(false);
-  const [selectedCustomPanels, setselectedCustomPanels] = useState<CustomPanelType[]>([]);
+  const [selectedCustomPanels, setselectedCustomPanels$] = useState<CustomPanelType[]>([]);
```
Comment (ps48) [inline public/components/custom_panels/custom_panel_table.tsx:94]: Not sure if the `$` is expected here. I see the below reference is `setselectedCustomPanels`

File: public/components/custom_panels/custom_panel_table.tsx [hunk 4]
```diff
@@ -137,41 +135,51 @@ export const CustomPanelTable = ({
...
-      console.log(err);
+        await dispatch(createPanel(newPanel));
+
+        setToast(`Observability Dashboard "${newName}" successfully created!`);
+      } catch (err) {
+        setToast(
+          'Error cloning Operational Panel, please make sure you have the correct permission.',
+          'danger'
+        );
+        console.log(err);
```
Comment (ps48) [inline public/components/custom_panels/custom_panel_table.tsx:165]: We should replace `console.log` with `console.error` to show errors.

File: public/components/custom_panels/custom_panel_table.tsx [hunk 5]
```diff
@@ -126,8 +120,12 @@ export const CustomPanelTable = ({
...
 
   const onCreate = async (newCustomPanelName: string) => {
-    const newPanel = newPanelTemplate(newCustomPanelName);
-    dispatch(createPanel(newPanel));
+    if (!isNameValid(newCustomPanelName)) {
+      setToast('Invalid Dashboard name', 'danger');
+    } else {
+      const newPanel = newPanelTemplate(newCustomPanelName);
+      dispatch(createPanel(newPanel));
+    }
```
[the remark this answers is not in the document]
Comment (derek-ho) [reply public/components/custom_panels/custom_panel_table.tsx:128]: I think we should be redirecting - that is the current behavior?

File: public/components/common/toast/index.tsx (hunk 2 above)
Comment (pjfitzgibbons) [reply public/components/common/toast/index.tsx:13]: Side is used when toast during flyout.   So unless toast on right-side is OK while visualization flyout and others are active, only then could we remove it.

Side is being passed-through on OSD API

File: public/components/custom_panels/custom_panel_table.tsx (hunk 3 above)
Comment (pjfitzgibbons) [reply public/components/custom_panels/custom_panel_table.tsx:94]: There was debugging.  Will double-check.

File: public/components/custom_panels/__tests__/custom_panel_view.test.tsx [hunk 6]
```diff
@@ -111,27 +134,31 @@ describe('Panels View Component', () => {
       window.location.assign(`#/event_analytics/explorer/${savedVisId}`);
     };
 
+    const services = { toasts: { addDanger: (t) => {} } }
```
Comment (joshuali925) [inline public/components/custom_panels/__tests__/custom_panel_view.test.tsx:137]: you can use https://github.com/opensearch-project/opensearch-dashboards/blob/main/src/core/TESTING.md#core-mocks

i added a mock here https://github.com/opensearch-project/dashboards-observability/blob/a4cb6d1d33d00349449932f4679144acba963ace/test/__mocks__/coreMocks.ts#L11

File: public/framework/redux/reducers/toast_slice.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { createSlice } from '@reduxjs/toolkit';
```
Comment (joshuali925) [inline public/framework/redux/reducers/toast_slice.ts:1]: license header missing

File: public/plugin.ts [hunk 8]
```diff
@@ -137,6 +138,10 @@ export class ObservabilityPlugin
       const savedObjects = new SavedObjects(coreStart.http);
       const timestampUtils = new TimestampUtils(dslService, pplService);
 
+      const services: ObservabilityAppServices = {
+        toasts: coreStart.notifications.toasts,
+      };
```
Comment (joshuali925) [inline public/plugin.ts:143]: why is this another service and not using your `coreRefs`?

File: public/components/common/toast/index.tsx (hunk 2 above)
Comment (ps48) [reply public/components/common/toast/index.tsx:13]: I may be missing something here. Can you please explain how are you using/passing down the side parameter in this function?

**Pull request closed without merging**